compiler/src/Language/Bond/Codegen/Java/Enum_java.hs

```diff
@@ -30,37 +30,73 @@ package #{javaPackage};
-        if (!(other instanceof #{declName})) { return false; }
-        return this.value == ((#{declName}) other).value;
-    }
+    public boolean equals(Object other) { return (other instanceof #{declName}) && (this.value == ((#{declName}) other).value); }
+ 
+    @Override
+    public int hashCode() { return this.value; }
 
     @Override
-    public int hashCode() {
-        return this.value;
+    public String toString() { return String.valueOf(this.value); }
+
+    public static #{declName} get(int value) {
+        switch (value) {
+            #{newlineSep 3 switchCaseConstantMapping enumConstantsWithIntDistinct}
+            default: return new #{declName}(value);
+        }
     }
 }|]
       where
-        -- constant
-        constant Constant {..} = let value x = [lt|#{x}|] in
-            [lt|public static final #{declName} #{constantName} = new #{declName}(#{optional value constantValue});|]
+        -- constant object
+        constantObjectDecl c@Constant {..} =
+            [lt|public static final #{declName} #{constantName} = #{enumObjectAssigmentValue c};|]
 
-        -- Process constants to make sure all values can be converted
-        -- to integer.
-        enumConstantsWithInt = fixEnumWithInt 0 enumConstants []
+        -- constant int
+        constantIntValueDecl Constant {..} = let value x = [lt|#{x}|] in
+            [lt|public static final int #{constantName} = #{optional value constantValue};|]
+ 
+        -- switch cases that map int to object
+        switchCaseConstantMapping Constant {..} =
+            [lt|case Values.#{constantName}: return #{constantName};|]
 
+        -- Process constants to make sure every constant value is set (either explicit or auto-generated)
+        enumConstantsWithInt = fixEnumWithInt 0 enumConstants []
+ 
         fixEnumWithInt :: Int -> [Constant] -> [Constant] -> [Constant]
         fixEnumWithInt _ [] result = reverse result
         fixEnumWithInt nextInt (h:r) result = case constantValue h of
           Just n -> fixEnumWithInt (n + 1) r (h:result)
           _ -> fixEnumWithInt (nextInt + 1) r ((Constant (constantName h) (Just nextInt)):result)
 
+        -- Process constants to make sure every constant value is distinct
+        enumConstantsWithIntDistinct = findEnumConstantsDistinct enumConstantsWithInt []
+
+        findEnumConstantsDistinct :: [Constant] -> [Maybe Int] -> [Constant]
+        findEnumConstantsDistinct [] _ = []
+        findEnumConstantsDistinct (h:r) keys = if elem (constantValue h) keys
+            then findEnumConstantsDistinct r keys
+            else h : findEnumConstantsDistinct r ((constantValue h):keys)
+
+        -- The RHS value to assign to an enum object
+        enumObjectAssigmentValue :: Constant -> Text
+        enumObjectAssigmentValue enumConstant =
+            let firstDeclaredEnumConstant = head (filter (\c -> (constantValue c) == (constantValue enumConstant)) enumConstantsWithIntDistinct) in
```

Comment:
This might be better expressed with [`any`](http://hackage.haskell.org/packages/archive/base/latest/doc/html/Prelude.html#v:any).